### PR TITLE
Implement rosout log widget.

### PR DESCRIPTION
**Pull request by MichaelGrupp**
Buffers formatted rosout messages in a ring buffer and displays them in
a box. Minimum loglevel, screen percentage and more are configurable.

In addition to enabling/disabling with 'l', I also added an intermediate
stopped state between enabled and disabled that can be useful to inspect
the last logs with a stopped buffer.

Known issues: log coloring sometimes doesn't work correctly, need to
investigate why (tui Text is weird). In the worst case you get white
text, so this is not a blocker atm.

### CODE REVIEW

File: src/app.rs [hunk 1]
```diff
@@ -334,5 +366,29 @@ impl App {
...
+        } else if self.rosout_widget_enabled {
+            self.rosout_widget_enabled = !self.rosout_widget_enabled;
+        } else {
+            self.rosout_listener.toggle_buffering();
+            self.rosout_widget_enabled = true;
+        }
+    }
+
+    pub fn build_rosout_widget(&mut self) -> Paragraph {
+        let logstrings = self.rosout_listener.read_logstring_buffer();
```
Comment (MaxiMaerz) [inline src/app.rs:386]: I think you can skip this declaration, you use it only once

File: src/rosout.rs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    min_loglevel: i8,
+) -> Option<String> {
+    use colored::{Color, Colorize};
+    use rosrust_msg::rosgraph_msgs::Log;
+
+    if level < min_loglevel {
+        return None;
+    }
+
+    let format_string = |levelname, color| {
```
Comment (MaxiMaerz) [inline src/rosout.rs:57]: fancy

File: src/rosout.rs [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+impl RosoutListener {
+    pub fn new(buffer_size: usize, start_buffering: bool, min_loglevel: i8) -> RosoutListener {
+        let log_buffer = Arc::new(RwLock::new(LogBuffer::new(buffer_size, start_buffering)));
+        let cb_log_buffer = log_buffer.clone();
+
+        let _subscriber = rosrust::subscribe(
+            "/rosout",
+            buffer_size,
+            move |rosout_msg: rosrust_msg::rosgraph_msgs::Log| {
```
Comment (MaxiMaerz) [inline src/rosout.rs:84]: I think we can leave it like this, but I find callback definitions like this not very clear

File: src/app.rs [hunk 4]
```diff
@@ -254,9 +271,24 @@ impl App {
...
+            let robot_view_percentage = 100 - self.rosout_screen_percentage;
+            chunks = Layout::default()
+                .constraints(
+                    [
+                        Constraint::Percentage(robot_view_percentage),
+                        Constraint::Percentage(self.rosout_screen_percentage),
+                    ]
+                    .as_ref(),
+                )
+                .split(f.size());
```
Comment (MaxiMaerz) [inline src/app.rs:286]: You need to change the size factors here to reflect the 80%
        let scale_factor = self.terminal_size.0 as f64 / self.terminal_size.1 as f64 * 0.5;
Here you need to add another multiplier

**Review by MaxiMaerz — CHANGES_REQUESTED** [review]
I tested it works nice!
I do not like capital L as a trigger, lower case is also fine

**Comment by MichaelGrupp** [comment]
Now it's 'l' to save your precious time. Warning: not to be confused with 'I'
https://www.youtube.com/watch?v=E3p_Crhi3K8

File: src/rosout.rs (hunk 3 above)
Comment (MichaelGrupp) [reply src/rosout.rs:84]: What would be your suggested alternative?

Comment (MaxiMaerz) [reply src/rosout.rs:84]: lets keep it

File: src/app.rs (hunk 4 above)
Comment (MichaelGrupp) [reply src/app.rs:292]: Done. The y-scaling had to changed: https://github.com/carzum/termviz/pull/43/commits/27e1c3cf0914bc86f46aed6a112ae983a0cf7ac4

File: src/app.rs [hunk 5]
```diff
@@ -334,5 +366,29 @@ impl App {
...
+            self.rosout_listener.toggle_buffering();
+        } else if self.rosout_widget_enabled {
+            self.rosout_widget_enabled = !self.rosout_widget_enabled;
+        } else {
+            self.rosout_listener.toggle_buffering();
+            self.rosout_widget_enabled = true;
+        }
+    }
+
+    pub fn build_rosout_widget(&mut self) -> Paragraph {
```
[the remark this answers is not in the document]
Comment (MichaelGrupp) [reply src/app.rs:391]: The rosout listener / buffer is accessed a member. I would keep the widget logic here to keep the rosout.rs free from GUI dependencies.

File: src/config.rs [hunk 6]
```diff
@@ -40,6 +40,25 @@ impl Default for TeleopConfig {
...
+    pub enabled_by_default: bool,
+}
+
+impl Default for RosoutConfig {
+    fn default() -> RosoutConfig {
+        RosoutConfig {
+            buffer_size: 150,
+            min_loglevel: 2, // INFO, see rosmsg show rosgraph_msgs/Log
+            screen_percentage: 25,
+            enabled_by_default: true,
```
Comment (carzum) [inline src/config.rs:57]: I would disable it by default, that will allow a quick glance at the robot on the complete terminal. Enabling it on demand with l is quick enough. But that might just be me. Any thumb up thumb downs for this?

File: src/rosout.rs (hunk 2 above)
Comment (carzum) [reply src/rosout.rs:56]: using 1.58 formatting for more fancyness

File: src/rosout.rs [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    let format_string = |levelname, color| {
+        format!("[{} {} {}] {}", name, levelname, stamp.seconds(), msg).color(color)
+    };
+
+    match level {
+        Log::DEBUG => Some(format!("{}", format_string("DEBUG", Color::White))),
+        Log::INFO => Some(format!("{}", format_string("INFO", Color::White))),
+        Log::WARN => Some(format!("{}", format_string("WARN", Color::Yellow))),
+        Log::ERROR => Some(format!("{}", format_string("ERROR", Color::Red))),
+        Log::FATAL => Some(format!("{}", format_string("FATAL", Color::Red))),
```
Comment (carzum) [inline src/rosout.rs:65]: fatal is bold afaik

File: src/app.rs [hunk 8]
```diff
@@ -334,5 +372,28 @@ impl App {
...
+        let mut all_spans = Vec::<Spans>::new();
+        for logstring in self.rosout_listener.read_logstring_buffer().iter() {
+            all_spans.extend(ansi_to_text(logstring.bytes()).unwrap().lines);
```
Comment (carzum) [inline src/app.rs:394]: the coloring is not consistent for me.  Why not store the message + level as a struct  in the buffer and use the level set  to set the foreground color on the span?  That would also remove the dependency

File: src/app.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (carzum) [reply src/app.rs:391]: The app.rs module is pretty bloated, adding more and more widget code in there wont make it better. I think your change is pretty minimal and ok, but i see this exploding on the long run

File: src/rosout.rs [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+        if self.producer.remaining() == 0 {
+            self.consumer.pop();
+        }
+        if self.producer.push(logstring).is_err() {
+            eprintln!("failed to buffer rosout log")
+        }
+    }
+
+    pub fn read_logstrings(&self) -> std::vec::Vec<String> {
+        let mut logstrings = std::vec::Vec::<String>::new();
```
Comment (Eman7C7) [inline src/rosout.rs:37]: this code is ugly, but I assume it's because Consumer doesn't support iterators right? :(

**Review by Eman7C7 — COMMENTED** [review]
I confirm that the colors are sometimes incorrect, but more importantly, At least from a quick test it misses some messages! Anyone with the same problem? (I'll test it better when I have more time)

**Review by MaxiMaerz — APPROVED** [review]

**Pull request closed without merging**